### ISSUE TITLE
How to get the experiment result on Occlusion LineMOD (LM-O) dataset ?

### ISSUE THREAD

**Issue by ZJU-PLP** [Gorilla-Lab-SCUT/BiCo-Net#2]
@CavendyHsu Hi, zelin

Cloud youd mind sharing the details to get  experiment result on Occlusion LineMOD (LM-O) dataset ? I want to reproduce the experiment result from your paper while cannot find the details in your released code.

**Comment by TerenceHsu666** [comment]
@ZJU-PLP Thank you for your interest in our work. We have updated the code and pre-trained models for the LMO dataset.

**Comment by ZJU-PLP** [comment]
@CavendyHsu Ok, Thanks for your reply.

I meet the problem when I want to download the Pred mask of HybridPose on LineMOD Occlusion dataset.

![image](https://user-images.githubusercontent.com/32757213/187899411-a4d80ff6-de30-45d2-8c39-1a83d09a2fee.png)

![image](https://user-images.githubusercontent.com/32757213/187899254-0f44980e-d2f9-4058-b641-1bc4545e0fea.png)

**Comment by TerenceHsu666** [comment]
Access is turned on.

**Comment by ZJU-PLP** [comment]
@CavendyHsu The file can be downloaded now. Thanks a lot.

**Comment by ZJU-PLP** [comment]
@CavendyHsu @aragakiyui611 Hi, Thanks for sharing the code to train and eval lmo dataset. However, I cannot reproduce your paper's results. Could you mind offering some advice?

1. I run the script and pretrained model is provided by your download [link](https://drive.google.com/drive/folders/18n506Vww8NipgWYxkVH0Cl2YIeufRZXz?usp=sharing). 

```python eval_lmo.py --dataset_root ./datasets/linemod/Linemod_preprocessed --pred_mask ./mask_hybridpose --model ./pretrained_models/lmo/pose_model_27_0.02377143515188488.pth```

However, I get the ADD(S) Result: **65.91** instead of **69.5** in your paper.
![image](https://user-images.githubusercontent.com/32757213/191157862-74216802-ea25-44ea-a392-eeb896c12677.png)

![image](https://user-images.githubusercontent.com/32757213/191160143-bbb67094-5352-44c6-84b6-976d8a601873.png)


2. I run the lmo dataset train script offered by your method. However, I cannot reproduce the good experiment results.(ADD(S) Result is only 25.15 and the loss cannot be refined)

```python train.py --dataset lmo --dataset_root path_to_lm_dataset --bg_img path_to_voc2012_dataset```

![image](https://user-images.githubusercontent.com/32757213/191159563-c35a95de-1991-4acd-8bca-ba5d2d4ab75a.png)
![image](https://user-images.githubusercontent.com/32757213/191159623-c782bb05-9259-4956-93c2-a0285d9203dd.png)
![image](https://user-images.githubusercontent.com/32757213/191160642-54be42a5-af3b-486d-b0a8-84ab2778424a.png)

**Comment by yunnan66** [comment]
@ZJU-PLP Hi.
When I was running the LMO dataset, I encountered the following problem. Do you have a solution? I am looking forward to your reply.
![截图 2022-09-20 17-25-24(1)](https://user-images.githubusercontent.com/90179991/191225085-3cc8efc9-1c69-4a37-ba25-d06904964268.png)

**Comment by ZJU-PLP** [comment]
> @ZJU-PLP Hi. When I was running the LMO dataset, I encountered the following problem. Do you have a solution? I am looking forward to your reply. ![截图 2022-09-20 17-25-24(1)](https://user-images.githubusercontent.com/90179991/191225085-3cc8efc9-1c69-4a37-ba25-d06904964268.png)

@yunnan66 
No, I did not meet the problem. Maybe your torchvision version is not supported (from your debug error infromation)?

**Comment by aragakiyui611** [comment]
<img width="729" alt="image" src="https://user-images.githubusercontent.com/74336517/192581259-453705e2-f22f-4eba-8130-d350ed22a820.png">
Hi, I tested the official weights of LMO and it is exactly 69.52, I am using torch1.8.0. Please check whether is the testing data problem or torch version

**Comment by ZJU-PLP** [comment]
@aragakiyui611 I have tested again (pytorch version changed to 1.8) and the result is still 65.62. Could you mind sharing your script bash command ?
![image](https://user-images.githubusercontent.com/32757213/192762442-0c61d4dc-6382-4cbc-b8f9-310432f3d98c.png)
![image](https://user-images.githubusercontent.com/32757213/192762865-ca3a030a-5e79-45a1-97d0-a67f88e58d20.png)

**Comment by aragakiyui611** [comment]
<img width="923" alt="image" src="https://user-images.githubusercontent.com/74336517/192765321-6dce8c73-2da2-4177-b126-a028491a8033.png">
<img width="489" alt="image" src="https://user-images.githubusercontent.com/74336517/192765358-409e611c-81de-4e3d-a582-9029c6ca873f.png">
<img width="299" alt="image" src="https://user-images.githubusercontent.com/74336517/192765785-7da7e659-7b3b-4711-b02a-ba4d98b3f856.png">

Please refer to these. I just simply run 
`python eval_lmo.py`

Pray check any modification on the dataset or network code.
Meanwhile I notice that it spent almost 24 min to test on lmo on your machine, while I only spent 8 min, I use 3090. Similarly,  your training time is quite long. I guess the bottleneck may be system IO instead of GPU.

**Comment by ZJU-PLP** [comment]
@aragakiyui611 
I have tested the code by "python eval_lmo.py --dataset_root ./datasets/linemod/Linemod_preprocessed --pred_mask ./mask_hybridpose --model ./pretrained_models/lmo/pose_model_27_0.02377143515188488.pth" and results is 65.78 in my lab server again (instead of my local desktop computer). The GPU is GTX 2080Ti.

I think your analysis is very right about system IO instead of GPU. However, I cannot understand  why the same model produce different results. Because system IO only influence running time while not the precision results. **I want to know whether you test the code in Linemod_preprocessed dataset?**
![image](https://user-images.githubusercontent.com/32757213/192790497-24694500-7ca2-4122-9b0b-9ed8f67cc528.png)

**Comment by aragakiyui611** [comment]
> I want to know whether you test the code in Linemod_preprocessed dataset.

Not quite understand, the 02 of Linemod_preprocessed is exactly Occluded LineMOD.

> However, I cannot understand why the same model produce different results.

This is very normal and there are so many factors affecting. You could probably search zhihu or others for detailed explanation.

As far as the result you run 65.78 and my 69.52, I can't find the exact where the problem is. Pray check whether you modified you test data.

**Comment by ZJU-PLP** [comment]
@aragakiyui611 
Hi, I have checked again LineMod dataset and found some error. Lately, I have fixed it and  tested again your provided trained model and the results is ok (dataset moved to SSD).
![image](https://user-images.githubusercontent.com/32757213/192945246-70ed3061-e830-416c-a64b-72e9be9cf0ca.png)

Today, I have trained lmo dataset again in my local computer. However, I find that the loss cannot be refined after epoch=25(**training --nepoch=50**). And the **test ADD(S) results is only 31.89**(model epoch=25).Could you mins sharing your **training curve displayed with tensorboard**?

Training details:
script command `python train.py --dataset lmo --dataset_root ./datasets/linemod/Linemod_preprocessed --bg_img ./datasets/lmo/VOCdevkit`
![image](https://user-images.githubusercontent.com/32757213/192946060-929add5d-0001-4930-ac59-95ae356102af.png)

![image](https://user-images.githubusercontent.com/32757213/192946071-e94eaddb-81da-4ba6-a05c-47f12659689a.png)

Testing details:
![image](https://user-images.githubusercontent.com/32757213/192946127-d4bb42cb-ac21-476b-9b97-389b8f02cac1.png)

![image](https://user-images.githubusercontent.com/32757213/192946153-d72a66ad-4dba-4c06-93ce-ecccb6ef6861.png)

**Comment by ZJU-PLP** [comment]
@aragakiyui611 Could you mind helping me to solve this problem when you are leisure?

**Comment by aragakiyui611** [comment]
@ZJU-PLP 
[exp.zip](https://github.com/Gorilla-Lab-SCUT/BiCo-Net/files/9741334/exp.zip)
These are training logs. If you need weights I may share it with other methods since they are too large.
I am afraid you have to figure out problem yourself first, I have not time in the recent month.

**Comment by ZJU-PLP** [comment]
@aragakiyui611 Thanks for your nice sharing. I find that my TEST FINISH Avg dis is different from yours to a great extent. I'll keep on debug the results.

**Comment by ZJU-PLP** [comment]
@aragakiyui611 I maybe find the key of this problem.

In your provided trained logs, the test result is different by **bgaug=True-front_num=3-num_pts=1000-box_aug=False** and **bgaug=True-front_num=3-num_pts=1000-box_aug=True**. From your provided trained model, the model of pose_model_27_0.02377143515188488.pth is trained by box_aug=True.

Could you mind helping me to understand the parameter of **box_aug**? In your lmo dataset training script from README, I cannot find the related parameter. 

epoch_27_test_log.txt:
(1) box_aug=False
Avg dis: **0.03193115701012664**
![image](https://user-images.githubusercontent.com/32757213/195337725-7a0944e1-f257-45aa-a2d1-0f89e978a5c2.png)
(2) box_aug=True
Avg dis: **0.02377143515188488**
![image](https://user-images.githubusercontent.com/32757213/195338040-06a3a957-c2a4-48de-8b3b-b5c418db21a3.png)

**Comment by aragakiyui611** [comment]
it makes background augmentation.
please refer to these lines:
background aug:

https://github.com/Gorilla-Lab-SCUT/BiCo-Net/blob/68bc075673f6696c0e4f18a67c5cd9be06900312/datasets/lmo/dataset.py#L105

bbox aug:

https://github.com/Gorilla-Lab-SCUT/BiCo-Net/blob/68bc075673f6696c0e4f18a67c5cd9be06900312/datasets/lmo/dataset.py#L129

**Comment by ZJU-PLP** [comment]
> it makes augmentation on detection bbox. please refer to these lines:
> 
> https://github.com/Gorilla-Lab-SCUT/BiCo-Net/blob/68bc075673f6696c0e4f18a67c5cd9be06900312/datasets/lmo/dataset.py#L129

Ok, I see.
While the parameters of  bgaug=True-front_num=3-num_pts=1000 stand for (bgaug, front_num and num_pts)?

**Comment by aragakiyui611** [comment]
I made a mistake, please refer to the edited comment.
`bgaug=True-front_num=3-num_pts=1000` is default.

**Comment by ZJU-PLP** [comment]
> I made a mistake, please refer to the edited comment. `bgaug=True-front_num=3-num_pts=1000` is default.

In your meanings, I can ignore the parameters(bgaug, front_num and num_pts) when to train lmo dataset?

**Comment by ZJU-PLP** [comment]
@aragakiyui611 
I trained several times in lmo dataset by your code. However, my train results cannot achieve your provided model. Could you mind helping me to train your published code again in lmo dataset by script `python train.py --dataset lmo --dataset_root path_to_lm_dataset --bg_img path_to_voc2012_dataset` (If it is convenient for you)? I am afraid that the published code maybe miss some key part.
 In other words, I can make sure that the problem location if you train **[the published code](https://github.com/Gorilla-Lab-SCUT/BiCo-Net)** (not your initial code)and can achieve the result **pose_model_27_0.02377143515188488.pth**.

My training results:
![image](https://user-images.githubusercontent.com/32757213/195509021-ba052395-811c-4eef-bf61-a58b8fc10774.png)
![image](https://user-images.githubusercontent.com/32757213/195509081-7bbb2361-82cb-409c-81b2-1c32a8fbbba9.png)

**Comment by aragakiyui611** [comment]
https://github.com/Gorilla-Lab-SCUT/BiCo-Net/blob/68bc075673f6696c0e4f18a67c5cd9be06900312/datasets/lmo/dataset.py#L37

I found that it may be this bug, I have updated the code

**Comment by ZJU-PLP** [comment]
> https://github.com/Gorilla-Lab-SCUT/BiCo-Net/blob/68bc075673f6696c0e4f18a67c5cd9be06900312/datasets/lmo/dataset.py#L37
> 
> I found that it may be this bug, I have updated the code

Ok, thanks a lot.
I am training again now.

**Comment by aragakiyui611** [comment]
<img width="232" alt="image" src="https://user-images.githubusercontent.com/74336517/195975097-751cb8ab-eda5-4ab6-b6e0-86aba7ebdd24.png">
I trained the model as well and there're still some bug.

**Comment by ZJU-PLP** [comment]
> <img alt="image" width="232" src="https://user-images.githubusercontent.com/74336517/195975097-751cb8ab-eda5-4ab6-b6e0-86aba7ebdd24.png">
> 
> I trained the model as well and there're still some bug.

I also trained again. It still exists some bugs.

![image](https://user-images.githubusercontent.com/32757213/196029936-af3d7b00-a65d-4a61-b1ec-e279b420fda7.png)
![image](https://user-images.githubusercontent.com/32757213/196030299-ff2a0bed-cd41-4e8a-aafb-4c0301d5ab58.png)

**Comment by ZJU-PLP** [comment]
@aragakiyui611 Could you mind helping me to solve this problem in lmo dataset when you are leisure? Maybe you can look up the initial version of experiment code which tested by the one of yours.

**Comment by ZJU-PLP** [comment]
@TerenceHsu666 Could you mind helping me to solve this problem in lmo dataset when you are leisure?

**Comment by aragakiyui611** [comment]
https://github.com/Gorilla-Lab-SCUT/BiCo-Net/blob/581671b554d5b33a17d69c45bd13d841cb0790e8/datasets/lmo/dataset.py#L130

Hi, please try change this line into `img_masked = np.transpose(img_masked, (2, 0, 1))`. 
I am training the model again.
It seems that this is the bug.

**Comment by ZJU-PLP** [comment]
> https://github.com/Gorilla-Lab-SCUT/BiCo-Net/blob/581671b554d5b33a17d69c45bd13d841cb0790e8/datasets/lmo/dataset.py#L130
> 
> Hi, please try change this line into `img_masked = np.transpose(img_masked, (2, 0, 1))`. I am training the model again. It seems that this is the bug.

Okkk, thanks a lot.

**Comment by ZJU-PLP** [comment]
@aragakiyui611 
I find the bug is fixed by your proposed method. The testing results is ok. Thanks for your patient response.
![image](https://user-images.githubusercontent.com/32757213/199038120-526179c1-932b-46cf-bb58-912ebab8563c.png)
![image](https://user-images.githubusercontent.com/32757213/199038155-6f5b7b02-08ad-4946-b490-03d3e0d50754.png)
